src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -437,86 +490,59 @@ protected internal void OnSolutionReloaded(SolutionInfo reloadedSolutionInfo)
+                WorkspaceChangeKind.ProjectAdded, projectId: projectInfo.Id);
         }
 
         /// <summary>
         /// Call this method to respond to a project being reloaded in the host environment.
         /// </summary>
         protected internal virtual void OnProjectReloaded(ProjectInfo reloadedProjectInfo)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                var projectId = reloadedProjectInfo.Id;
-
-                CheckProjectIsInCurrentSolution(projectId);
-
-                var oldSolution = this.CurrentSolution;
-                var newSolution = oldSolution.RemoveProject(projectId).AddProject(reloadedProjectInfo);
-
-                newSolution = this.AdjustReloadedProject(oldSolution.GetRequiredProject(projectId), newSolution.GetRequiredProject(projectId)).Solution;
-                newSolution = this.SetCurrentSolution(newSolution);
+            var projectId = reloadedProjectInfo.Id;
+            this.SetCurrentSolution(
+                oldSolution =>
+                {
+                    CheckProjectIsInSolution(oldSolution, projectId);
 
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectReloaded, oldSolution, newSolution, projectId);
-            }
+                    return this.AdjustReloadedProject(
+                        oldSolution.GetRequiredProject(projectId),
+                        oldSolution.RemoveProject(projectId).AddProject(reloadedProjectInfo).GetRequiredProject(projectId)).Solution;
+                }, WorkspaceChangeKind.ProjectReloaded, projectId);
         }
 
         /// <summary>
         /// Call this method to respond to a project being removed from the host environment.
         /// </summary>
         protected internal virtual void OnProjectRemoved(ProjectId projectId)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                CheckProjectIsInCurrentSolution(projectId);
-                this.CheckProjectCanBeRemoved(projectId);
-
-                var oldSolution = this.CurrentSolution;
-
-                this.ClearProjectData(projectId);
-                var newSolution = this.SetCurrentSolution(oldSolution.RemoveProject(projectId));
+            this.SetCurrentSolution(
+                oldSolution =>
+                {
+                    CheckProjectIsInSolution(oldSolution, projectId);
+                    this.CheckProjectCanBeRemoved(projectId);
 
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectRemoved, oldSolution, newSolution, projectId);
-            }
+                    return oldSolution.RemoveProject(projectId);
+                },
+                WorkspaceChangeKind.ProjectRemoved, projectId,
+                onBeforeUpdate: (oldSolution, _) =>
+                {
+                    // Clear out mutable state not associated with teh solution snapshot (for example, which documents are
```

Comment:
Does this being before vs. after actually matter? No reason to do it in this PR but I wonder if we can merge the two.

Reply:
(I later saw we have the DocumentClosing callbacks that demanded a "before"...I wonder if we can just delete those.)

Reply:
yeah, this part wasn't clear to me.  so i wanted to preserve teh semantics that we update the mutable workspace state prior to updating latestSolution.  it would be nice if we can simplify and do afterwards.  it will need more care to make sure that the cleanup code doesn't access .CurrentSolution.